comm/sync.go

```diff
@@ -57,66 +70,70 @@ func fetchBlocks(ctx context.Context, peer *Peer, fromBlockNum uint32, fetched c
 			return nil
 		}
 
-		blocks := make([]*block.Block, 0, len(result))
-		for _, raw := range result {
-			var blk block.Block
-			if err := rlp.DecodeBytes(raw, &blk); err != nil {
-				return errors.Wrap(err, "invalid block")
-			}
-			if blk.Header().Number() != fromBlockNum {
-				return errors.New("broken sequence")
-			}
-			fromBlockNum++
-			blocks = append(blocks, &blk)
+		batch := rawBlockBatch{
+			rawBlocks: result,
+			startNum:  fromBlockNum,
 		}
 
 		select {
-		case fetched <- blocks:
+		case rawBatches <- batch:
 		case <-ctx.Done():
 			return ctx.Err()
 		}
+
+		fromBlockNum += uint32(len(result))
 	}
 }
 
-func warmupBlocks(ctx context.Context, fetched <-chan []*block.Block, warmedUp chan<- *block.Block) {
-	<-co.Parallel(func(queue chan<- func()) {
-		for blocks := range fetched {
-			for _, blk := range blocks {
-				queue <- func() {
-					blk.Header().ID()
-					blk.Header().Beta()
-				}
-				for _, tx := range blk.Transactions() {
-					queue <- func() {
-						tx.ID()
-						tx.UnprovedWork()
-						_, _ = tx.IntrinsicGas()
-						_, _ = tx.Delegator()
-					}
-				}
-			}
+func decodeAndWarmupBatches(ctx context.Context, rawBatches <-chan rawBlockBatch, warmedUp chan<- *block.Block) error {
+	g, ctx := errgroup.WithContext(ctx)
 
-			for _, blk := range blocks {
-				select {
-				case <-ctx.Done():
-					return
-				case warmedUp <- blk:
-				}
-
-				// when queued blocks count > 10% warmed up channel cap,
-				// send nil block to throttle to reduce mem pressure.
-				if len(warmedUp)*10 > cap(warmedUp) {
-					const targetSize = 2048
-					for range int(blk.Size())/targetSize - 1 {
-						select {
-						case warmedUp <- nil:
-						default:
-						}
-					}
-				}
-			}
+	for batch := range rawBatches {
+		g.Go(func() error {
+			return decodeAndWarmupBatch(ctx, batch, warmedUp)
+		})
+	}
```

Comment:
This might send out of order blocks to the `warmedUp chan` which will trigger a broken sequence error.